### PR TITLE
Portfolio update emails

### PR DESCRIPTION
**Pull request by IanPhilips**
Screen Shot 2022-09-23 at 4 05 38 PM

### CODE REVIEW

File: functions/src/weekly-portfolio-emails.ts [hunk 1]
```diff
@@ -0,0 +1,280 @@
...
+  log(
+    'Sending weekly portfolio emails to',
+    privateUsersToSendEmailsTo.length,
+    'users'
+  )
+
+  const usersBets: { [userId: string]: Bet[] } = {}
+  // get all bets made by each user
+  await Promise.all(
+    privateUsersToSendEmailsTo.map(async (user) => {
```
[the remark this answers is not in the document]
Comment (mantikoros) [reply functions/src/weekly-portfolio-emails.ts:64]: Maybe abstract the logic for a single user into a function sendPortfolioUpdateEmail() and then process batches of 200 users in parallel.

File: functions/src/email-templates/weekly-portfolio-update.html [hunk 2]
```diff
@@ -0,0 +1,510 @@
...
+                                        </td>
+                                      </tr>
+                                        <td align="center"
+                                            style="font-size:0px;padding:10px 20px;padding-top:0px;padding-bottom:0px;word-break:break-word;">
+                                            <table border="0" cellpadding="0" cellspacing="0" role="presentation"
+                                                style="border-collapse:collapse;border-spacing:0px; ">
+                                                <tbody>
+                                                    <tr>
+                                                      <th style='width: 170px'>
+                                                        🔥 Prediction streak
```
[the remark this answers is not in the document]
Comment (IanPhilips) [reply functions/src/email-templates/weekly-portfolio-update.html:268]: Consistent with something other than the notifications ui?
Screen Shot 2022-09-29 at 4 31 54 PM

File: functions/src/weekly-portfolio-emails.ts [hunk 3]
```diff
@@ -0,0 +1,280 @@
...
+  // get all users that haven't unsubscribed from weekly emails
+  const privateUsersToSendEmailsTo = privateUsers
+    .filter((user) => {
+      return isProd()
+        ? user.notificationPreferences.profit_loss_updates.includes('email') &&
+            !user.weeklyPortfolioUpdateEmailSent
+        : true
+    })
+    // Send emails in batches
+    .slice(0, 200)
```
[the remark this answers is not in the document]
Comment (IanPhilips) [reply functions/src/weekly-portfolio-emails.ts:54]: It gets the first 200 emails every minute for an hour.

File: functions/src/weekly-portfolio-emails.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (IanPhilips) [reply functions/src/weekly-portfolio-emails.ts:64]: I'm following similar logic to what the weekly markets emails does and it works great.

File: functions/src/email-templates/weekly-portfolio-update.html (hunk 2 above)
[the remark this answers is not in the document]
Comment (mantikoros) [reply functions/src/email-templates/weekly-portfolio-update.html:268]: Ah, I thought I changed that back to "trading"...

**Pull request merged**